### DISCUSSION TITLE
[API Proposal]: Create `fullnameof(element)` expression

### DISCUSSION THREAD

**Discussion by DrkWzrd** [dotnet/csharplang#5200] [Language Ideas] [answered]
### Background and motivation

We have an excellent timesaving and error avoidance expression called `nameof(ElementToGetName)` so useful in so many cases.

In some other cases (usually logging-related events, debugging, even some errors validations) developers need the full name of a class with the namespace or the full name of a member with the class name.

The first thing require a hard-coded string (which you'll have to update every time you rename the class or the namespace, or even if you modify the hierarchy of namespaces) or a smelly `typeof(Class).FullName` which has a performance penalty if called so many times.
And the second thing requires a concatenation of `nameof(Class)+"."+nameof(Member)`.

### API Proposal

An expression like said: 
```C#
fullnameof(ElementToGetName)
```     

### API Usage

```C#
// A example class
namespace Store.Shelf.Box { class Potato { int Quality {get;} } }
// A example call;
var box_fullName = fullnameof(Box); //It returns "Store.Shelf.Box"
var potato_fullName = fullnameof(Potato); //It returns "Store.Shelf.Box.Potato"
var quality_fullName = fullnameof(Potato.Quality); //It returns "Potato.Quality", right now nameof(Potato.Quality) returns "Quality"

```     

### Risks

At first sight, no one i can see.

**Comment by yaakov-h** [comment, marked as answer]
This is quite similar to and a partial duplicate of #701.

**Comment by jnm2** [comment]
I'm proposing making `typeof(`...`).FullName` a constant, which will help with some of these use cases: https://github.com/dotnet/csharplang/issues/8505

It would be great to hear in that thread about real-world use cases that support doing this, as well as real-world use cases for `fullnameof` that _aren't_ addressed by that proposal and don't have a good solution today.

**Reply by jehrenzweig-captionhealth** [reply]
https://github.com/dotnet/csharplang/issues/8505 has been marked "_Closed as not planned_".
